### PR TITLE
Improve failure modes for os.rename() as used in distribution caching.

### LINKED ISSUE #265: Failure to extract dependencies to ~/.pex/install if they already exist with some Python version
Specifically on a host running Python 2.7.5 we ran into the following stack trace:

```
Traceback (most recent call last):
  File ".bootstrap/_pex/pex.py", line 309, in execute
  File ".bootstrap/_pex/pex.py", line 78, in _activate
  File ".bootstrap/_pex/environment.py", line 132, in activate
  File ".bootstrap/_pex/environment.py", line 176, in _activate
  File ".bootstrap/_pex/environment.py", line 121, in update_candidate_distributions
  File ".bootstrap/_pex/environment.py", line 107, in load_internal_cache
  File ".bootstrap/_pex/environment.py", line 95, in write_zipped_internal_cache
  File ".bootstrap/_pex/util.py", line 177, in cache_distribution
OSError: [Errno 17] File exists
```

In this case we're on a slightly older version of pex, line 177 in util.py is: https://github.com/pantsbuild/pex/blob/ba58a854d561dcf6963c0097cad544bfe2861c5a/pex/util.py#L177

So, the rename fails because the directory exists. In this case it's raising EEXIST, whereas pex only expects ENOTEMPTY to be a continuable error code.

Pex should handle EEXIST with the same behavior as ENOTEMPTY in this case.

### PR DESCRIPTION
**Pull request by kwlzn**
Fixes #265.

### CODE REVIEW

File: pex/common.py [hunk 1]
```diff
@@ -140,6 +140,21 @@ def safe_rmtree(directory):
...
+def rename_if_empty(src, dest, allowable_errors=(errno.EEXIST, errno.ENOTEMPTY)):
+  """Rename `src` to `dest` using `os.rename()`.
+
+  If an `OSError` with errno in `allowable_errors` is encountered during the rename, this is
+  a functional no-op and the src directory will be removed.
+  """
+  try:
+    os.rename(src, dest)
+  except OSError as e:
+    if e.errno in allowable_errors:
```
Comment (peiyuwang) [inline pex/common.py]: in the case of `ENOTEMPTY`, it's no-op only if `dest == src`, otherwise we remove `src` but `dest != src`

Comment (kwlzn) [inline pex/common.py]: correct - it's not pretty, but this is the current behavior.

Comment (kwlzn) [inline pex/common.py]: eliminated the `no-op` terminology.

Comment (peiyuwang) [inline pex/common.py]: :shipit:

**Pull request merged**